File: src/main/java/io/github/suitetecsa/sdk/nauta/ConnectClient.java
package io.github.suitetecsa.sdk.nauta;

import io.github.suitetecsa.sdk.nauta.exception.*;
import io.github.suitetecsa.sdk.nauta.model.ConnectInformation;
import io.github.suitetecsa.sdk.nauta.model.DataSession;
import io.github.suitetecsa.sdk.nauta.network.Action;
import io.github.suitetecsa.sdk.nauta.network.HttpMethod;
import io.github.suitetecsa.sdk.nauta.network.PortalCommunicator;
import io.github.suitetecsa.sdk.nauta.network.action.*;
import io.github.suitetecsa.sdk.nauta.scraper.ConnectionInfoParser;
import io.github.suitetecsa.sdk.nauta.scraper.ErrorParser;
import io.github.suitetecsa.sdk.nauta.scraper.FormParser;
import io.github.suitetecsa.sdk.nauta.utils.Constants;
import io.github.suitetecsa.sdk.nauta.utils.ExceptionHandler;
import io.github.suitetecsa.sdk.nauta.utils.PortalManager;
import io.github.suitetecsa.sdk.nauta.utils.StringUtils;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.util.AbstractMap.SimpleEntry;
import java.util.List;
import java.util.Map;

public class ConnectClient {

    private static String @NotNull [] init(@NotNull PortalCommunicator communicator) throws NautaAttributeException, NautaException, LoadInfoException {
        FormParser scraper = new FormParser.Builder().build();

        SimpleEntry<String, Map<String, String>> loginResponse = communicator.performRequest(new GetPage("https://" + Constants.connectDomain + ":8443", null, HttpMethod.GET), scraper::parseLoginForm);
        String loginAction = loginResponse.getKey();
        Map<String, String> loginData = loginResponse.getValue();

        String wlanUserIp = loginData.getOrDefault("wlanuserip", "");
        String csrfHW = loginData.getOrDefault("CSRFHW", "");

        return new String[]{loginAction, wlanUserIp, csrfHW};
    }

    public static Boolean isConnected() throws NautaAttributeException, NautaException, LoadInfoException {
        PortalCommunicator communicator = new PortalCommunicator.Builder().build();
        ConnectionInfoParser scraper = new ConnectionInfoParser.Builder().build();
        return communicator.performRequest(new CheckConnection(null), httpResponse -> scraper.parseCheckConnection(httpResponse.getText()));
    }

    public static Long getRemainingTime(DataSession dataSession) throws NautaAttributeException, NautaException, LoadInfoException {
        if (!isConnected()) {
            throw new LoadInfoException("you are not connected. You need to be connected to get the remaining accountTime.");
        }

        PortalCommunicator communicator = new PortalCommunicator.Builder().build();

        Action action = new LoadUserInformation(dataSession.username(), null, dataSession.wlanUserIp(), dataSession.csrfHw(), dataSession.attributeUUID(), PortalManager.CONNECT);
        return communicator.performRequest(action, httpResponse -> {
            try {
                return StringUtils.toSeconds(httpResponse.getText());
            } catch (InvalidSessionException e) {
                throw new RuntimeException(e);
            }
        });
    }

    public static ConnectInformation getUserInformation(String username, String password) throws NautaAttributeException, NautaException, LoadInfoException {
        PortalCommunicator communicator = new PortalCommunicator.Builder().build();
        ConnectionInfoParser scraper = new ConnectionInfoParser.Builder().build();

        String[] initResult = init(communicator);
        String wlanUserIp = initResult[1];
        String csrfHw = initResult[2];

        Action action = new LoadUserInformation(username, password, wlanUserIp, csrfHw, null, PortalManager.CONNECT);
        return communicator.performRequest(action, httpResponse -> {
            try {
                return scraper.parseConnectInformation(httpResponse);
            } catch (NautaException | LoadInfoException e) {
                throw new RuntimeException(e);
            }
        });
    }

    @Contract("_, _ -> new")
    public static @NotNull DataSession connect(String username, String password) throws NautaAttributeException, NautaException, LoadInfoException, LoginException {
        ExceptionHandler<LoginException> loginExceptionHandler = new ExceptionHandler.Builder<LoginException>().build();
        if (isConnected()) {
            throw loginExceptionHandler.handleException("Fail to connect", List.of("Already connected"));
        }

        PortalCommunicator communicator = new PortalCommunicator.Builder().build();
        ConnectionInfoParser scraper = new ConnectionInfoParser.Builder().build();

        String[] initResult = init(communicator);
        String wlanUserIp = initResult[1];
        String csrfHw = initResult[2];

        Action action = new Login(csrfHw, wlanUserIp, username, password, password, PortalManager.CONNECT, HttpMethod.POST);
        String uuid = communicator.performRequest(action, httpResponse -> {
            try {
                new ErrorParser.Builder().whenPortalManager(PortalManager.CONNECT).build().throwExceptionOnFailure(
                        httpResponse,
                        "Fail to login",
                        loginExceptionHandler
                );
                return scraper.parseAttributeUUID(httpResponse);
            } catch (LoadInfoException | LoginException | NotLoggedInException e) {
                throw new RuntimeException(e);
            }
        });
        return new DataSession(username, csrfHw, wlanUserIp, uuid);
    }

    public static void disconnect(DataSession dataSession) throws NautaAttributeException, NautaException, LoadInfoException, LogoutException {
        ExceptionHandler<LogoutException> logoutExceptionHandler = new ExceptionHandler.Builder<LogoutException>().build();
        if (!isConnected()) {
            throw logoutExceptionHandler.handleException("Fail to disconnect", List.of("You are not connected"));
        }

        PortalCommunicator communicator = new PortalCommunicator.Builder().build();
        ConnectionInfoParser scraper = new ConnectionInfoParser.Builder().build();

        Action action = new Logout(dataSession.username(), dataSession.wlanUserIp(), dataSession.csrfHw(), dataSession.attributeUUID());
        Boolean isLogout = communicator.performRequest(action, scraper::isSuccessLogout);
        if (!isLogout) throw logoutExceptionHandler.handleException("Failed to disconnect", List.of(""));
    }
}

File: src/main/kotlin/io/github/suitetecsa/sdk/nauta/ConnectClient.kt
package io.github.suitetecsa.sdk.nauta

import io.github.suitetecsa.sdk.nauta.exception.*
import io.github.suitetecsa.sdk.nauta.model.ConnectInformation
import io.github.suitetecsa.sdk.nauta.model.DataSession
import io.github.suitetecsa.sdk.nauta.network.*
import io.github.suitetecsa.sdk.nauta.network.ConnectRoute.*
import io.github.suitetecsa.sdk.nauta.scraper.ConnectionInfoParser
import io.github.suitetecsa.sdk.nauta.scraper.ErrorParser
import io.github.suitetecsa.sdk.nauta.scraper.FormParser
import io.github.suitetecsa.sdk.nauta.utils.*
import org.jetbrains.annotations.Contract

object ConnectClient {
    @Throws(NautaAttributeException::class, NautaException::class, LoadInfoException::class)
    private fun init(communicator: PortalCommunicator): Array<String> {
        val scraper = FormParser.Builder().build()

        val loginResponse = communicator.performRequest(Init, scraper::parseLoginForm)
        val loginAction = loginResponse.key
        val loginData = loginResponse.value

        val wlanUserIp = loginData.getOrDefault("wlanuserip", "")
        val csrfHW = loginData.getOrDefault("CSRFHW", "")

        return arrayOf(loginAction, wlanUserIp, csrfHW)
    }

    /**
     * Verifica si hay conexión a internet
     *
     * @return true si hay conexión, false en caso contrario
     * @throws NautaAttributeException sí hay un error obteniendo los atributos
     * @throws NautaException sí hay un error de conexión general
     * @throws LoadInfoException sí hay un error cargando la información
     */
    @JvmStatic
    @get:Throws(NautaAttributeException::class, NautaException::class, LoadInfoException::class)
    val isConnected: Boolean
        get() {
            val communicator = PortalCommunicator.Builder().build()
            val scraper = ConnectionInfoParser.Builder().build()
            return communicator.performRequest(CheckAccess, scraper::parseCheckConnection)
        }

    /**
     * Obtiene el tiempo restante de la sesión
     *
     * @param dataSession datos de la sesión
     * @return el tiempo restante en segundos
     * @throws NautaAttributeException si hay error obteniendo los atributos
     * @throws NautaException si hay error de conexión
     * @throws LoadInfoException si no hay conexión
     */
    @JvmStatic
    @Throws(NautaAttributeException::class, NautaException::class, LoadInfoException::class)
    fun getRemainingTime(dataSession: DataSession): Long {
        if (!isConnected) {
            throw LoadInfoException("you are not connected. You need to be connected to get the remaining accountTime.")
        }

        val communicator = PortalCommunicator.Builder().build()

        val route: ConnectRoute = GetTime(
            username = dataSession.username,
            wlanUserIp = dataSession.wlanUserIp,
            csrfHw = dataSession.csrfHw,
            attributeUUID = dataSession.attributeUUID
        )
        return communicator.performRequest(route) { StringUtils.toSeconds(it.text) }
    }

    /**
     * Obtiene la información de conexión del usuario
     *
     * @param username el nombre de usuario
     * @param password la contraseña
     * @return la información de conexión
     * @throws NautaAttributeException si hay error obteniendo atributos
     * @throws NautaException si hay error de conexión
     * @throws LoadInfoException sí hay error cargando la información
     */
    @JvmStatic
    @Throws(NautaAttributeException::class, NautaException::class, LoadInfoException::class)
    fun getUserInformation(username: String, password: String): ConnectInformation {
        val communicator = PortalCommunicator.Builder().build()
        val scraper = ConnectionInfoParser.Builder().build()

        val initResult = init(communicator)
        val wlanUserIp = initResult[1]
        val csrfHw = initResult[2]

        val route: ConnectRoute = UserInfo(
            username = username,
            password = password,
            wlanUserIp = wlanUserIp,
            csrfHw = csrfHw
        )
        return communicator.performRequest(route, scraper::parseConnectInformation)
    }

    /**
     * Realiza la conexión a internet con las credenciales dadas
     *
     * @param username el nombre de usuario
     * @param password la contraseña
     * @return la sesión de datos de la conexión
     * @throws NautaAttributeException si hay error obteniendo atributos
     * @throws NautaException si hay error de conexión
     * @throws LoadInfoException sí hay error cargando información
     * @throws LoginException si falla el login
     */
    @JvmStatic
    @Contract("_, _ -> new")
    @Throws(NautaAttributeException::class, NautaException::class, LoadInfoException::class, LoginException::class)
    fun connect(username: String, password: String): DataSession {
        val loginExceptionHandler = ExceptionHandler { LoginException(it) }
        if (isConnected) {
            throw loginExceptionHandler.handleException("Fail to connect", listOf("Already connected"))
        }

        val communicator = PortalCommunicator.Builder().build()
        val scraper = ConnectionInfoParser.Builder().build()

        val initResult = init(communicator)
        val wlanUserIp = initResult[1]
        val csrfHw = initResult[2]

        val route: ConnectRoute =
            Connect(username = username, password = password, csrfHw = csrfHw, wlanUserIp = wlanUserIp)
        val uuid = communicator.performRequest(route) { httpResponse: HttpResponse ->
            ErrorParser.Builder().whenPortalManager(PortalManager.CONNECT).build().throwExceptionOnFailure(
                httpResponse,
                "Fail to login",
                loginExceptionHandler
            )
            scraper.parseAttributeUUID(httpResponse)
        }
        return DataSession(username, csrfHw, wlanUserIp, uuid)
    }

    /**
     * Desconecta la sesión actual
     *
     * @param dataSession datos de la sesión a desconectar
     * @throws NautaAttributeException si hay error obteniendo atributos
     * @throws NautaException si hay error de conexión
     * @throws LoadInfoException si no hay conexión
     * @throws LogoutException si falla el logout
     */
    @JvmStatic
    @Throws(NautaAttributeException::class, NautaException::class, LoadInfoException::class, LogoutException::class)
    fun disconnect(dataSession: DataSession) {
        val logoutExceptionHandler = ExceptionHandler { LogoutException(it) }
        if (!isConnected) {
            throw logoutExceptionHandler.handleException("Fail to disconnect", listOf("You are not connected"))
        }

        val communicator = PortalCommunicator.Builder().build()
        val scraper = ConnectionInfoParser.Builder().build()

        val route: ConnectRoute = Disconnect(
            username = dataSession.username,
            wlanUserIp = dataSession.wlanUserIp,
            csrfHw = dataSession.csrfHw,
            attributeUUID = dataSession.attributeUUID
        )
        val isLogout = communicator.performRequest(route, scraper::isSuccessLogout)
        if (!isLogout) throw logoutExceptionHandler.handleException("Failed to disconnect", listOf(""))
    }
}
